runtime/ms_rest_azure/lib/ms_rest_azure/credentials/msi_token_provider.rb

```diff
@@ -78,12 +78,68 @@ def initialize(port = 50342, settings = ActiveDirectoryServiceSettings.get_azure
     #
     # @return [String] authentication headers.
     def get_authentication_header
-      acquire_token if token_expired
+      if !ENV['MSI_VM'].nil? && ENV['MSI_VM'].downcase == 'true'
```

Comment:
Wonder what is this environment variable of `MSI_VM` used for? Also is there a in-memory token cache in place to mitigate the throttling issue?

Reply:
what is this environment variable of MSI_VM used for? 

If the variable is set to true, then the token is obtained using the old method. 

 is there a in-memory token cache in place to mitigate the throttling issue?

Actually Yes. We do not get a new token until the current one is expired.

Reply:
> Actually Yes. We do not get a new token until the current one is expired.

Token is per resource and per identity. If the same provider could be used by multiple clients or multiple identities, then a single token entry might not be enough. I am not familiar with how the `MSITokenProvider` gets used, but hope it helps.

Reply:
That helps. Actually the same provider will not be used in the Ruby SDK. All of them use their own instance and works fine for all the other token providers.

Reply:
@sarangan12 , sorry for not calling out earlier, but if you have not done it, can you set the `user-agent` header to token requests send to imds. This will help AAD group to use telemetry to diagnose throttling symptoms